SPEC.md

```diff
@@ -0,0 +1,334 @@
+
+- `202 Accepted`: Cancelation request accepted.
+
+  **Body**: Unspecified.
+
+- `404 Not Found`: Operation ID not recognized or references deleted.
+
+  **Body**: Unspecified.
+
+### Get Operation Result
+
+Retrieve operation result.
+
+#### Method
+
+`GET`
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}/{operation_id}/result`
+
+#### Query Parameters
+
+- `wait_deadline`: Optional. ISO 8601 date indicating the waiting period for a result.
+  If not specified, and the operation is still running, the request should resolve immediately with the current
+  operation status.
+
+#### Response Codes
+
+- `200 OK`: Operation completed successfully.
+
+  **Headers**:
+
+  - `Nexus-Operation-State: succeeded`
+
+  **Body**: Arbitrary data conveying the operation's result. Headers should specify content type and encoding.
+
+- `204 No Content`: Operation completed successfully with an empty result or is still running.
+
+  **Headers**:
+
+  - `Nexus-Operation-State: succeeded | running`
+
+  **Body**: Empty.
+
+- `482 Operation Failed`: Operation completed as `failed` or `canceled`.
+
+  **Headers**:
+
+  - `Content-Type: application/json`
+  - `Nexus-Operation-State: failed | canceled`
+
+  **Body**: A JSON serialized [`Failure`](#failure) object.
+
+- `408 Request Timeout`: The specified `wait_deadline` expired prior to operation completion.
+
+  The caller may re-issue this request to start a new long poll.
+
+  **Body**: Unspecified.
+
+- `404 Not Found`: Operation ID not recognized or references deleted.
+
+  **Body**: Unspecified.
+
+### Get Operation Info
+
+Retrieve operation details.
+
+#### Method
+
+`GET`
+
+#### Paths
+
+- `/api/v1/services/{service}/operations/{operation}/{operation_id}`
+
+#### Response Codes
+
+- `200 OK`: Successfully retrieved info.
+
+  **Headers**:
+
+  - `Content-Type: application/json`
+  - `Nexus-Operation-State: failed | canceled`
+
+  **Body**:
+
+  A JSON serialized [`OperationInfo`](#operationinfo) object.
+
+- `404 Not Found`: Operation ID not recognized or references deleted.
+
+## General Information on HTTP Response Codes
+
+The Nexus protocol follows standard HTTP practices:
+
+- `4xx` codes signify non-retryable framework-level errors.
+- `5xx` codes signify retryable framework-level errors.
+
+## Callback URLs
+
+Any HTTP URL can be used to deliver operation completions.
+
+Callers should ensure URLs contain sufficient information to correlate completions with initiators.
+
+For invoking a callback URL:
+
+- Include the `Nexus-Operation-State` header.
+- If state is `succeeded`, deliver non-empty results in the body with corresponding `Content-*` headers.
+- If state is `failed` or `canceled`, content type should be `application/json` and the body must have a serialized [`Failure`](#failure) object.
+- Upon successful completion delivery, the handler should reply with a `200 OK` status and an empty body.
+
+### Security
+
+There's no enforced security for callback URLs.
```

Comment:
I think some callers might be in a position where they need auth tokens set in headers for the callback request; how can they specify the required headers?

Reply:
Not currently specified as part of the spec and we're encouraging tokens in the URL. However, specific implementations such as the Temporal server may support per endpoint/service configuration for outbound Nexus calls, like headers. Auth can get complex quick (what if they want mTLS? what if they want server-oauth? etc) and is often implementation specific, so we want to avoid a spec requiring something specific at this time.

Reply:
> so we want to avoid a spec requiring something specific at this time.

I'm not suggesting requiring something -- it's the opposite; I'm suggesting making it more flexible by allowing them to supply headers required for auth.